runtime/pkg/middleware/request.go

```diff
@@ -0,0 +1,105 @@
+package middleware
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"github.com/rilldata/rill/runtime/pkg/observability"
+	"google.golang.org/grpc"
+)
+
+// This is a collection of gRPC and HTTP interceptors that call fn per request.
+
+// RequestStreamServerInterceptor wraps a ServerStream and calls fn on each RecvMsg.
+func RequestStreamServerInterceptor(fn func(Metadata) error) grpc.StreamServerInterceptor {
+	return func(
+		srv interface{},
+		ss grpc.ServerStream,
+		info *grpc.StreamServerInfo,
+		handler grpc.StreamHandler,
+	) error {
+		_, method, err := parseFullMethod(info.FullMethod)
+		if err != nil {
+			return err
+		}
+
+		wss := wrappedServerStream{ss, method, observability.GrpcPeer(ss.Context()), fn}
+		return handler(srv, &wss)
+	}
+}
+
+// RequestUnaryServerInterceptor calls fn on each request
+func RequestUnaryServerInterceptor(fn func(Metadata) error) grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
+		_, method, err := parseFullMethod(info.FullMethod)
+		if err != nil {
+			return nil, err
+		}
+
+		md := Metadata{ctx, req, method, observability.GrpcPeer(ctx)}
+		if err := fn(md); err != nil {
+			return nil, err
+		}
+
+		return handler(ctx, req)
+	}
+}
```

Comment:
Consider replacing these with `grpc_auth.UnaryServerInterceptor`, like here: https://github.com/rilldata/rill/blob/199453658f812528fb5c2ec7e87f599809e9ac2c/admin/server/server.go#L129C20-L129C20

Thoughts:

- You can use `grpc.Method(ctx)` to get the full method and `observability.GrpcPeer(ctx)` to get the peer.
- When rate limiting on a per-method basis, we should probably use the full method instead of parsing it, since the service prefix matters (we have multiple services, and they might use the same Redis).
- (It might still make sense to create a `grpcutil.ParseMethod(ctx)` func or something to get the individual service and method name)
- (Using the full method has another benefit, which is that we can use the generated [constants here](https://github.com/rilldata/rill/blob/199453658f812528fb5c2ec7e87f599809e9ac2c/proto/gen/rill/admin/v1/api_grpc.pb.go#L21) for the method names if we need to do manual overrides. The timeout middleware currently doesn't use it – but it probably should!)

Reply:
Agree, updated to use the method